pgx/src/datum/range.rs

```diff
@@ -0,0 +1,391 @@
+/*
+Portions Copyright 2019-2021 ZomboDB, LLC.
+Portions Copyright 2021-2022 Technology Concepts & Design, Inc. <support@tcdi.com>
+
+All rights reserved.
+
+Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+*/
+
+//! Utility functions for working with `pg_sys::RangeType` structs
+use crate::{
+    pg_sys, void_mut_ptr, Date, FromDatum, IntoDatum, Numeric, Timestamp, TimestampWithTimeZone,
+};
+use pgx_pg_sys::{Oid, RangeBound};
+use pgx_utils::sql_entity_graph::metadata::{
+    ArgumentError, Returns, ReturnsError, SqlMapping, SqlTranslatable,
+};
+use std::marker::PhantomData;
+
+/// Represents Datum to serialized RangeType PG struct
+pub struct Range<T: FromDatum + IntoDatum + RangeSubType> {
+    ptr: *mut pg_sys::varlena,
+    range_type: *mut pg_sys::RangeType,
+    _marker: PhantomData<T>,
+}
+
+impl<T> Range<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    /// ## Safety
+    /// function requires that
+    /// - datum is null OR
+    /// - datum represents a RangeType datum
+    #[inline]
+    unsafe fn from_pg(ptr: *mut pg_sys::varlena, range_type: *mut pg_sys::RangeType) -> Self {
+        Range {
+            ptr,
+            range_type,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<T> TryFrom<pg_sys::Datum> for Range<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    type Error = RangeConversionError;
+
+    /// ## Safety
+    /// function requires that
+    /// - datum is null OR
+    /// - datum represents a RangeType datum
+    fn try_from(datum: pg_sys::Datum) -> Result<Self, Self::Error> {
+        if datum.is_null() {
+            Err(RangeConversionError::NullDatum)
+        } else {
+            unsafe {
+                let ptr = datum.ptr_cast();
+                let range_type =
+                    pg_sys::pg_detoast_datum(datum.ptr_cast()) as *mut pg_sys::RangeType;
+                let _ = range_type.as_ref().expect("RangeType * was NULL");
```

Comment:
Is there a reason not to return an Err here (and expand the RangeConversionError to include this)?

Reply:
...Could the returned pointer be unaligned?

Reply:
I removed this vestigial check.